price-estimator/src/solver_rounding_buffer.rs

```diff
@@ -16,45 +16,50 @@ fn max_rounding_amount(token_price: f64, fee_token_price: f64) -> f64 {
     max_rounding_amount.max(1.0)
 }
 
+/// Calculate a single rounding buffer like the solver does. This amount is subtracted from the
+/// denominator of orders selling the sell token and buying the buy token.
+pub fn rounding_buffer(fee_token_price: f64, sell_token_price: f64, buy_token_price: f64) -> f64 {
+    let estimated_xrate = buy_token_price / sell_token_price;
+    max_rounding_amount(buy_token_price, fee_token_price)
+        * estimated_xrate
+        * PRICE_ESTIMATION_ERROR.powi(2)
+}
+
 /// Perform the same rounding buffer calculation as our solvers in order to increase the correctness
 /// of our estimates.
 /// All token prices must be nonzero.
 #[allow(dead_code)]
 pub fn apply_rounding_buffer(
-    tokens: &Tokens,
+    token_prices: &TokenPrices,
     orders: &mut Vec<Order>,
     account_state: &mut AccountState,
-    pessimistic_factor: f64,
+    extra_factor: f64,
 ) {
-    let fee_token_price = tokens
+    let fee_token_price = *token_prices
         .get(&TokenId(0))
-        .expect("fee token does not have price")
-        .external_price as f64;
+        .expect("fee token does not have price") as f64;
     assert!(fee_token_price > 0.0, "fee token price is 0");
 
     // The maximum rounding buffer over all orders from this address selling this token.
     let mut account_balance_buffers = HashMap::<(Address, TokenId), u128>::new();
     // Apply rounding buffer to account balances and order sell amounts.
     for order in orders.iter_mut() {
-        // Compute rounding buffer.
         // If a token doesn't have a price it means it isn't connected to the fee. In this case we
         // use a price of 1 to get some reasonable default rounding buffer.
         let (sell_token, buy_token) = (TokenId(order.sell_token), TokenId(order.buy_token));
-        let estimated_buy_token_price = match tokens.get(&buy_token) {
-            Some(TokenInfo { external_price, .. }) if *external_price > 0 => *external_price as f64,
+        let buy_token_price = match token_prices.get(&buy_token) {
+            Some(price) if *price > 0 => *price as f64,
```

Comment:
Should the default consider the token's decimals? Depending on if a higher price leads to a larger or lesser rounding buffer we should probably choose a price of `10^^(decimals +/- 5 or the like)`. This would allow to cover prices from 0,0001 to 10000

Reply:
Just saw #1137, let's discuss there.

Reply:
This is a good point. I've already changed this in another PR, so let's discuss it there.